components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/basicauth/BasicAuthCredentialValidator.java

```diff
@@ -308,46 +193,49 @@ public boolean validateScopes(String username, OpenAPI openAPI, MessageContext s
         String apiContext = (String) synCtx.getProperty(RESTConstants.REST_API_CONTEXT);
         String apiVersion = (String) synCtx.getProperty(RESTConstants.SYNAPSE_REST_API_VERSION);
         String apiElectedResource = (String) synCtx.getProperty(APIConstants.API_ELECTED_RESOURCE);
-
-        String[] userRoles = userRoleList;
+        String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
         org.apache.axis2.context.MessageContext axis2MessageContext = ((Axis2MessageContext) synCtx)
                 .getAxis2MessageContext();
         String httpMethod = (String) axis2MessageContext.getProperty(APIConstants.DigestAuthConstants.HTTP_METHOD);
         String resourceKey = apiContext + ":" + apiVersion + ":" + apiElectedResource + ":" + httpMethod;
+        Map<String, Scope> scopeMap = apiKeyValidator.retrieveScopes(tenantDomain);
         String resourceCacheKey = resourceKey + ":" + username;
 
         if (gatewayKeyCacheEnabled && getGatewayBasicAuthResourceCache().get(resourceCacheKey) != null) {
             return true;
         }
 
         if (openAPI != null) {
-            String resourceRoles = null;
             // retrieve the user roles related to the scope of the API resource
             List<String> resourceScopes = OpenAPIUtils.getScopesOfResource(openAPI, synCtx);
             List<String> resourceRolesList = new ArrayList<>();
             if (resourceScopes != null && resourceScopes.size() > 0) {
                 for (String resourceScope : resourceScopes) {
-                    String roles = OpenAPIUtils.getRolesOfScope(openAPI, synCtx, resourceScope);
-                    if (StringUtils.isNotEmpty(roles)) {
-                        resourceRolesList.add(roles);
+                    Scope scope = scopeMap.get(resourceScope);
+                    if (scope != null) {
+                        if (scope.getRoles().isEmpty()) {
+                            log.debug("Scope " + resourceScope + " Didn't have roles");
+                            return true;
```

Comment:
Is it necessary to add to gateway key cache before returning ?

Reply:
since this was a memory map no need to cache it

Reply:
(Y)